include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -109,1507 +390,1185 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
                    + (blockDim.x * blockDim.y) * threadIdx.z;
 
-    // initialize shared memory
-    T val = m_tally_device->tally.val;
-    Index_type idx = m_tally_device->tally.idx;
-    for (int i = BLOCK_SIZE / 2; i > 0; i /= 2) {
-      // this descends all the way to 1
-      if (threadId < i) {
-        sd_val[threadId + i] = val;
-        sd_idx[threadId + i] = idx;
-      }
-    }
-    if (threadId < 1) {
-      sd_val[threadId] = val;
-      sd_idx[threadId] = idx;
-    }
+    int warpId = threadId % WARP_SIZE;
+    int warpNum = threadId / WARP_SIZE;
 
-    __syncthreads();
-#else
-    m_is_copy_host = true;
-    m_smem_offset = getCudaSharedmemOffset(m_myID,
-                                           BLOCK_SIZE,
-                                           (sizeof(T) + sizeof(Index_type)));
-#endif
-  }
+    cuda::LocType<T, IndexType> temp;
+    temp.val = val;
+    temp.idx = idx;
 
-  /*!
-   * \brief Finish reduction on device and free memory on host.
-   *
-   * Destruction on host releases the global memory block chunk for
-   * reduction id and id itself for others to use.
-   * Destruction on device completes the reduction.
-   *
-   * Note: destructor executes on both host and device.
-   */
-  __host__ __device__ ~ReduceMaxLoc<cuda_reduce<BLOCK_SIZE, Async>, T>()
-  {
-#if defined(__CUDA_ARCH__)
-    if (m_finish_reduction) {
-      extern __shared__ unsigned char sd_block[];
-      T *sd_val = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
-      Index_type *sd_idx = reinterpret_cast<Index_type *>(
-          &sd_block[m_smem_offset + sizeof(T) * BLOCK_SIZE]);
+    if (numThreads % WARP_SIZE == 0) {
 
-      int blockId = blockIdx.x + blockIdx.y * gridDim.x
-                    + gridDim.x * gridDim.y * blockIdx.z;
-
-      int blocks = gridDim.x * gridDim.y * gridDim.z;
-
-      int threadId = threadIdx.x + blockDim.x * threadIdx.y
-                     + (blockDim.x * blockDim.y) * threadIdx.z;
+      // reduce each warp
+      for (int i = 1; i < WARP_SIZE ; i *= 2) {
+        T rhs_val = internal::shfl_xor_sync<T>(temp.val, i);
+        IndexType rhs_idx = internal::shfl_xor_sync<T>(temp.idx, i);
+        Reducer{}(temp.val, temp.idx, rhs_val, rhs_idx);
+      }
 
-      __syncthreads();
+    } else {
 
-      for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
-        if (threadId < i) {
-          RAJA_MAXLOC_UNSTRUCTURED(sd_val[threadId],
-                                   sd_idx[threadId],
-                                   sd_val[threadId],
-                                   sd_idx[threadId],
-                                   sd_val[threadId + i],
-                                   sd_idx[threadId + i]);
+      // reduce each warp
+      for (int i = 1; i < WARP_SIZE ; i *= 2) {
+        int srcLane = threadId ^ i;
+        T rhs_val = internal::shfl_sync<T>(temp.val, srcLane);
+        IndexType rhs_idx = internal::shfl_sync<T>(temp.idx, srcLane);
+        // only add from threads that exist (don
```

Comment:
You can just say ReduceMin and not create the self type alias. The fully qualified template list is implied.